presto-native-execution/presto_cpp/main/PeriodicMemoryChecker.h

```diff
@@ -0,0 +1,127 @@
+    /// the server memory usage is beyond this limit. This only applies if
+    /// 'systemMemPushbackEnabled_' is true.
+    uint64_t systemMemLimitBytes{0};
+
+    /// Specifies the amount of memory to shrink when the memory pushback is
+    /// triggered. This only applies if 'systemMemPushbackEnabled_' is true.
+    uint64_t systemMemShrinkBytes{0};
+
+    /// If true, starts memory limit checker that dumps the heap profile if
+    /// memory allocated via malloc exceeds 'mallocBytesUsageDumpThreshold_'.
+    /// See 'mallocCheckingCb()' for more details.
+    bool mallocMemHeapDumpEnabled{false};
+
+    /// Minimum Interval in seconds between two consecutive heap dumps.
+    size_t minHeapDumpIntervalSec{10};
+
+    /// The directory where heap profiles are dumped.
+    std::string heapDumpLogDir;
+
+    /// The prefix of heapdump file name.
+    std::string heapDumpFilePrefix;
+
+    /// Maximum number of heap dump files to maintain in rotation.
+    int maxHeapDumpFiles{10};
+
+    /// Memory (in bytes) allocated via malloc() that triggers the heap dump.
+    /// Default is 20GB.
+    size_t mallocBytesUsageDumpThreshold{20UL * 1024 * 1024 * 1024};
+  };
+
+  explicit PeriodicMemoryChecker(Config config);
+
+  void start();
+
+  void stop();
+
+ protected:
+  /// Returns current system memory usage. The returned value is used to compare
+  /// with 'Config::systemMemLimitBytes'.
+  virtual int64_t systemUsedMemoryBytes() const = 0;
+
+  /// Returns current bytes allocated by malloc. The returned value is used to
+  /// compare with 'Config::mallocBytesUsageDumpThreshold'
+  virtual int64_t mallocBytes() const = 0;
+
+  /// Callback function that is invoked by 'PeriodicMemoryChecker' at all time.
+  /// Light operations such as stats reporting can be done in this call back.
+  virtual void periodicCb() const = 0;
+
+  /// Callback function that performs a heap dump. Returns true if dump is
+  /// successful.
+  virtual bool heapDumpCb(const std::string& filePath) const = 0;
+
+  const Config config_;
+
+ private:
+  // Struct that stores the file names of the heap profiles dumped and the
+  // memory allocated by jemalloc when they were dumped. Used to create a min
+  // priority queue ordered by the memory allocated by jemalloc when the heap
+  // dump was generated. Used to determine which heap profile to delete when the
+  // number of files generated exceeds 'maxHeapDumpFiles_'.
+  struct DumpFileInfo {
+    // The memory allocated via jemalloc when the heap dump was generated.
+    int64_t mallocUsedBytes;
+    // Path to the heap dump file.
+    std::string filePath;
+
+    bool operator>(const DumpFileInfo& other) const {
+      return mallocUsedBytes > other.mallocUsedBytes;
+    }
+    bool operator<(const DumpFileInfo& other) const {
+      return mallocUsedBytes < other.mallocUsedBytes;
+    }
+  };
+
+  // Invoked by the periodic checker when 'Config::systemMemPushbackEnabled'
```

Comment:
What's specific callback that run this? Thanks!

Reply:
This is done in scheduler.